clients/src/main/java/org/apache/kafka/common/record/FileLogBuffer.java

```diff
@@ -0,0 +1,457 @@
+            Record shallowRecord = shallowEntry.record();
+            if (shallowRecord.timestamp() >= targetTimestamp) {
+                // We found a message
+                for (LogEntry deepLogEntry : shallowEntry) {
+                    long timestamp = deepLogEntry.record().timestamp();
+                    if (timestamp >= targetTimestamp)
+                        return new TimestampAndOffset(timestamp, deepLogEntry.offset());
+                }
+                throw new IllegalStateException(String.format("The message set (max timestamp = %s, max offset = %s" +
+                        " should contain target timestamp $targetTimestamp but it does not.", shallowRecord.timestamp(),
+                        shallowEntry.offset()));
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Return the largest timestamp of the messages after a given position in this file message set.
+     * @param startingPosition The starting position.
+     * @return The largest timestamp of the messages after the given position.
+     */
+    public TimestampAndOffset largestTimestampAfter(int startingPosition) {
+        long maxTimestamp = Record.NO_TIMESTAMP;
+        long offsetOfMaxTimestamp = -1L;
+
+        Iterator<FileChannelLogEntry> shallowIterator = shallowEntriesFrom(startingPosition);
+        while (shallowIterator.hasNext()) {
+            LogEntry shallowEntry = shallowIterator.next();
+            long timestamp = shallowEntry.record().timestamp();
+            if (timestamp > maxTimestamp) {
+                maxTimestamp = timestamp;
+                offsetOfMaxTimestamp = shallowEntry.offset();
+            }
+        }
+        return new TimestampAndOffset(maxTimestamp, offsetOfMaxTimestamp);
+    }
+
+    @Override
+    public boolean hasMatchingShallowMagic(byte magic) {
+        Iterator<FileChannelLogEntry> iterator = shallowEntriesFrom(Integer.MAX_VALUE, start, false);
+        while (iterator.hasNext())
+            if (iterator.next().magic() != magic)
+                return false;
+        return true;
+    }
+
+    @Override
+    public Iterator<FileChannelLogEntry> shallowEntries() {
+        return shallowEntriesFrom(start);
+    }
+
+    public Iterator<FileChannelLogEntry> shallowEntries(int maxMessageSize) {
+        return shallowEntriesFrom(maxMessageSize, start, true);
+    }
+
+    private Iterator<FileChannelLogEntry> shallowEntriesFrom(long start) {
+        return shallowEntriesFrom(Integer.MAX_VALUE, start, true);
+    }
+
+    private Iterator<FileChannelLogEntry> shallowEntriesFrom(int maxMessageSize, long start, boolean eagerLoadRecords) {
+        final long end;
+        if (isSlice)
+            end = this.end;
+        else
+            end = this.sizeInBytes();
+        FileLogInputStream inputStream = new FileLogInputStream(channel, maxMessageSize, start, end, eagerLoadRecords);
+        return LogBufferIterator.shallowIterator(inputStream);
```

Comment:
Also, in Java, having named methods is clearer than using booleans since one cannot use named arguments. However, it can be a bit confusing to have both options.